src/inversion/inversion.f90

```diff
@@ -440,82 +470,73 @@ end subroutine pressure
+            integer                         :: nc
 
-            ! calculate df1/dx
-            call fftxyp2s(f(:, :, :, I_X), fs)
-            call diffx(fs, ds)
-            call fftxys2p(ds, f(:, :, :, I_X))
+            do nc = 1, 3
+               call field_combine_semi_spectral(fs(:, :, :, nc))
+            enddo
 
-            !$omp parallel workshare
-            f(  -1, :, :, I_X) = two * f( 0, :, :, I_X) - f(   1, :, :, I_X)
-            f(nz+1, :, :, I_X) = two * f(nz, :, :, I_X) - f(nz-1, :, :, I_X)
-            !$omp end parallel workshare
+            ! calculate df1/dx
+            call diffx(fs(:, :, :, 1), divs)
 
             ! calculate df2/dy
-            call fftxyp2s(f(:, :, :, I_Y), fs)
-            call diffy(fs, ds)
-            call fftxys2p(ds, f(:, :, :, I_Y))
+            call diffy(fs(:, :, :, 2), ds)
 
-            !$omp parallel workshare
-            f(  -1, :, :, I_Y) = two * f( 0, :, :, I_Y) - f(   1, :, :, I_Y)
-            f(nz+1, :, :, I_Y) = two * f(nz, :, :, I_Y) - f(nz-1, :, :, I_Y)
-            !$omp end parallel workshare
+            divs = ds + divs
 
             ! calculate df3/dz
-            call central_diffz(f(:, :, :, I_Z), div)
+            call central_diffz(fs(:, :, :, 3), ds)
 
             ! div = df1/dx + df2/dy + df3/dz
-            div = f(:, :, :, I_X) + f(:, :, :, I_Y) + div
+            divs = ds + divs
+
+            do nc = 1, 3
+               call field_decompose_semi_spectral(fs(:, :, :, nc))
+            enddo
 
         end subroutine divergence
 
         !::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
 
         ! Computes a divergent flow field (ud, vd, wd) = grad(phi) where
         ! Lap(phi) = div (given) and grad = (ud, vd, wd).
-        subroutine diverge(div, grad)
-            double precision, intent(inout)  :: div(-1:nz+1, 0:ny-1, 0:nx-1)
-            double precision, intent(out)    :: grad(-1:nz+1, 0:ny-1, 0:nx-1, 3)
-            double precision                 :: ds(0:nz, 0:ny-1, 0:nx-1)
-            double precision                 :: us(0:nz, 0:ny-1, 0:nx-1), &
-                                                vs(0:nz, 0:ny-1, 0:nx-1), &
-                                                ws(0:nz, 0:ny-1, 0:nx-1)
-
-            !------------------------------------------------------------------
-            ! Convert phi to spectral space (in x & y) as ds:
-            call fftxyp2s(div, ds)
+        subroutine diverge(divs, grad)
+            double precision, intent(inout)  :: divs(0:nz, 0:nx-1, 0:ny-1)
+            double precision, intent(out)    :: grad(0:nz, 0:ny-1, 0:nx-1, 3)
+            double precision                 :: us(0:nz, 0:nx-1, 0:ny-1), &
+                                                vs(0:nz, 0:nx-1, 0:ny-1), &
+                                                ws(0:nz, 0:nx-1, 0:ny-1)
 
-            ds(:, 0, 0) = zero
+            divs(:, 0, 0) = zero
```

Comment:
I don't think we need this anymore if we are doing a 3D inversion.  In fact, when k = l = 0, we need to treat the problem differently and solve just for dphi/dz.  But to get zero boundary values, we need a condition that the integral of divs(iz, 0, 0) (discrete integral) is zero.  Seem a lot of trouble...